### PR TITLE
[#3196] Common request object for Flask and Pylons

### PR DESCRIPTION
**Pull request by amercader**
A new CKAN Request object based on Werkzeug's LocalProxy that will forward to Flask or Pylons own request objects depending on the output of `_get_request` (which essentially calls `is_flask_request`), and at the same time provide all objects methods to be able to interact with them transparently.

We don't use LocalProxy directly so we can handle the special case of accessing query string parameters via `request.params` on Flask (which uses `request.args`). All new Flask-only code should use `args` but for backwards compatibility we will support `params` for a while.

### CODE REVIEW

File: ckan/common.py [hunk 1]
```diff
@@ -25,6 +25,25 @@
...
+        pylons_request_available = True
+    except TypeError:
+        pylons_request_available = False
+
+    if (flask.request and
+            (flask.request.environ.get(u'ckan.app') == u'flask_app' or
+             not pylons_request_available)):
+        return True
+    else:
+        return False
```
Comment (wardi) [inline ckan/common.py]: you can `return (flask.req....` here, no need for an if/else :smile:

File: ckan/common.py [hunk 2]
```diff
@@ -91,10 +110,46 @@ def __delitem__(self, key):
...
+        All new code meant to be run just in Flask (eg views) should always
+        use request.args
+        '''
+        try:
+            return super(CKANRequest, self).__getattr__(name)
+        except AttributeError:
+            if name == u'params':
+                return super(CKANRequest, self).__getattr__(u'args')
+            else:
+                raise
```
Comment (wardi) [inline ckan/common.py]: would this work?
```python
    @property
    def params(self):
        try:
            return super(CKANRequest, self).params
        except AttributeError:
            return self.args
```
Maybe just code golf, but it seems less broad to me.

**Comment by amercader** [comment]
Please don't merge this until 2.6 is branched

**Comment by amercader** [comment]
Merged as part of #3212

**Pull request closed without merging**